### ISSUE TITLE
postgres connector - ssl connection not working

### ISSUE THREAD

**Issue by sebasaeros** [airbytehq/airbyte#17000]
## Environment
- **Airbyte version**:  0.40.7
- **OS Version / Instance**: Debian 10
- **Deployment**: Docker
- **Source Connector and version**: Postgresql 1.0.8
- **Step where error happened**:  Setup new connection 

## Current Behavior

When I specified all the parameters for the verify-ca connection mode in psql Airbyte give me an error as seen in the image

Also I cant use the require ssl connection mode because I dont have the field to complete with the key and crt.

## Expected Behavior

To connect in verify-ca mode or require mode.

As showed in the last image I cant provide the key and crt for require mode connection

## Steps to Reproduce
1. Create a new connector
2. Specify require mode ssl or verify-ca
3. Try to set up a connection
![airbyte](https://user-images.githubusercontent.com/104517224/191592957-0751041b-baa1-43c9-9db2-82bbcae63446.png)
![result](https://user-images.githubusercontent.com/104517224/191592960-8e48cee1-8665-43aa-bb4e-56d47183302b.png)
![image](https://user-images.githubusercontent.com/104517224/191593588-40363451-661a-43b6-9076-8f231d18e0f5.png)

**Comment by marcosmarxm** [comment]
This error is happening because a misconfiguration in your Postgres database. Check discussion here: https://stackoverflow.com/questions/67443028/why-am-i-getting-sql-state-28000-fatal-ident-authentication-failed-for-user-er

**Comment by sebasaeros** [comment]
But in require mode connection I need a certificate and Airbyte doesn't give me the option to pass it for the connection

**Comment by marcosmarxm** [comment]
@sebasaeros can you share the complete log file to identify the error message?

**Comment by henriblancke** [comment]
@rodireich I'm running into the same issue as @sebasaeros. I just tried removing the `------BEGIN RSA PRIVATE KEY-----` from the client key only and that gives me `Could not connect with provided configuration. Error: Failed to create keystore for Client certificate`. Let me know if there are more details I can provide?

**Comment by rodireich** [comment]
Thanks @henriblancke 
It looks like it didn't like the key or certificate provided.
Can you please share the full error log - with stacktrace.

**Comment by henriblancke** [comment]
@rodireich what is the easiest way to get the full error logs (including stacktraces) for new source checks? Should I just copy what the source check outputs in the airbyte UI?

**Comment by henriblancke** [comment]
@rodireich this is the full log output, it doesn't seem to contain any stacktraces:
```
2022-09-28 19:54:36 INFO i.a.w.t.TemporalAttemptExecution(get):106 - Docker volume job log path: /tmp/workspace/34a33f1e-8e95-4d8d-a8a1-829ead8aa7b5/0/logs.log
2022-09-28 19:54:36 INFO i.a.w.t.TemporalAttemptExecution(get):111 - Executing worker wrapper. Airbyte version: 0.40.9
2022-09-28 19:54:36 INFO i.a.c.i.LineGobbler(voidCall):114 - 
2022-09-28 19:54:36 INFO i.a.c.i.LineGobbler(voidCall):114 - ----- START CHECK -----
2022-09-28 19:54:36 INFO i.a.c.i.LineGobbler(voidCall):114 - 
2022-09-28 19:54:36 INFO i.a.c.i.LineGobbler(voidCall):114 - Checking if airbyte/source-postgres:1.0.11 exists...
2022-09-28 19:54:36 INFO i.a.c.i.LineGobbler(voidCall):114 - airbyte/source-postgres:1.0.11 was found locally.
2022-09-28 19:54:36 INFO i.a.w.p.DockerProcessFactory(create):119 - Creating docker container = source-postgres-check-34a33f1e-8e95-4d8d-a8a1-829ead8aa7b5-0-azcvg with resources io.airbyte.config.ResourceRequirements@51a6636a[cpuRequest=,cpuLimit=,memoryRequest=,memoryLimit=]
2022-09-28 19:54:36 INFO i.a.w.p.DockerProcessFactory(create):163 - Preparing command: docker run --rm --init -i -w /data/34a33f1e-8e95-4d8d-a8a1-829ead8aa7b5/0 --log-driver none --name source-postgres-check-34a33f1e-8e95-4d8d-a8a1-829ead8aa7b5-0-azcvg --network host -v airbyte_workspace:/data -v /tmp/airbyte_local:/local -e DEPLOYMENT_MODE=OSS -e USE_STREAM_CAPABLE_STATE=true -e WORKER_ENVIRONMENT=DOCKER -e AIRBYTE_ROLE= -e WORKER_JOB_ATTEMPT=0 -e WORKER_CONNECTOR_IMAGE=airbyte/source-postgres:1.0.11 -e AIRBYTE_VERSION=0.40.9 -e WORKER_JOB_ID=34a33f1e-8e95-4d8d-a8a1-829ead8aa7b5 airbyte/source-postgres:1.0.11 check --config source_config.json
2022-09-28 19:54:38 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Running source under deployment mode: OSS
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Starting source: io.airbyte.integrations.base.ssh.SshWrappedSource
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - integration args: {check=null, config=source_config.json}
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Running integration: io.airbyte.integrations.base.ssh.SshWrappedSource
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Command: CHECK
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Integration config: IntegrationConfig{command=CHECK, configPath='source_config.json', catalogPath='null', statePath='null'}
2022-09-28 19:54:39 WARN i.a.w.i.DefaultAirbyteStreamFactory(internalLog):100 - Unknown keyword order - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2022-09-28 19:54:39 WARN i.a.w.i.DefaultAirbyteStreamFactory(internalLog):100 - Unknown keyword airbyte_secret - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2022-09-28 19:54:39 WARN i.a.w.i.DefaultAirbyteStreamFactory(internalLog):100 - Unknown keyword multiline - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Starting connection with method: NO_TUNNEL
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - VERIFY_CA toSslJdbcParam verify-ca
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - HikariPool-1 - Starting...
2022-09-28 19:54:39 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - HikariPool-1 - Start completed.
2022-09-28 19:55:40 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - HikariPool-1 - Shutdown initiated...
2022-09-28 19:55:44 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - HikariPool-1 - Shutdown completed.
2022-09-28 19:55:44 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Completed integration: io.airbyte.integrations.base.ssh.SshWrappedSource
2022-09-28 19:55:44 INFO i.a.w.i.DefaultAirbyteStreamFactory(internalLog):103 - Completed source: io.airbyte.integrations.base.ssh.SshWrappedSource
2022-09-28 19:55:44 INFO i.a.w.t.TemporalAttemptExecution(get):132 - Stopping cancellation check scheduling...
2022-09-28 19:55:44 INFO i.a.c.i.LineGobbler(voidCall):114 - 
2022-09-28 19:55:44 INFO i.a.c.i.LineGobbler(voidCall):114 - ----- END CHECK -----
2022-09-28 19:55:44 INFO i.a.c.i.LineGobbler(voidCall):114 -
```

The UI completes with:

![image](https://user-images.githubusercontent.com/1708162/192877023-5ab6e8cb-e9dc-449f-a32f-10c90b1c397f.png)


The used CA Certificate, Client certificate and key here work correctly with other pg clients, I'm able to connect successfully. 

Thanks again for the help!

**Comment by rodireich** [comment]
Thank you @henriblancke 
I'm trying to see how to navigate to those logs. Any idea @marcosmarxm ?

Can I ask if you know how the certificates were created?
Do you have by any chance any sample you can share? 
The error indicates that when attempting to generate a client keystore from your client certificate and key, the operation failed.

**Comment by henriblancke** [comment]
@rodireich the certificates were created by an external vendor. The error in my initial message (`Could not connect with provided configuration. Error: Failed to create keystore for Client certificate`) was when I removed the cert header and footer. That did result in a failure to generate a keystore. But when I add the header and footer back it seems to be able to create the keystore but for some reason is not using the client certificate? I also don't see the `sslcert` param set on the jdbc connection url? It only appends `sslkey` and `sslrootcert`?

**Comment by rodireich** [comment]
I see. thanks for clarifying about removal of BEGIN KEY header.

Postgres is expecting `sslKey=<path to pkcs12 keystore>` 
Reading [postgres' docs](https://jdbc.postgresql.org/documentation/ssl/) it says

> As of version 42.2.9 PKCS-12 is also supported. In this archive format the client key and the client certificate are in one file, which needs to be set with the sslkey parameter. For the PKCS-12 format to be recognized, the file extension must be “.p12” (supported since 42.2.9) or “.pfx” (since 42.2.16). (In this case the sslcert parameter is ignored.)

Which indicates that sslcert is no longer needed (the pkcs12 keystore file packages both key and cert).

**Comment by rodireich** [comment]
> (supported since 42.2.9)
Could this be the issue? Do know which version of server you connection is done against?

**Comment by rodireich** [comment]
@henriblancke Was this a CDC (Logical replication) or a standard setup?

**Comment by henriblancke** [comment]
@rodireich a standard setup

**Comment by rodireich** [comment]
We're building with postgres jdbc 42.3.5 which is later than 42.2.9 so should be OK

**Comment by rodireich** [comment]
do you know which version of postgres this is connecting with?

**Comment by henriblancke** [comment]
yep :) `PostgreSQL 13.7 on x86_64-pc-linux-gnu, compiled by Debian clang version 12.0.1, 64-bit`

**Comment by rodireich** [comment]
Thanks. Which ssl mode is configured?
And also, the certificates provided by a vendor. do you know whether they are signed by a 3rd party CA - the likes of digicert or entrust ? or is it self-signed

**Comment by henriblancke** [comment]
@rodireich I tried both `verify-ca` and `verify-full` both didn't work. I'm not sure if the certs were signed by a 3rd party CA, it looks like the certs were issued by `Google Cloud SQL Client CA`. Hope that helps?

**Comment by rodireich** [comment]
It does help, thank you for the great info. 
One last question - is this an instance of [google cloud sql](https://cloud.google.com/sql/docs/postgres/db-versions)?
I'm trying to understand where the problem is, since until I get to logs it sounds like everything worked as was intended but connection fails.

**Comment by sebasaeros** [comment]
As I see we are having exactly the same problem, I can connect for any other client with the certificates but no from Airbyte

**Comment by rodireich** [comment]
I see the same error on Google cloud sql instance

**Comment by rodireich** [comment]
I'd like to understand why this isn't behaving according to doc. 
But please do. this is awesome!

**Comment by rodireich** [comment]
Hi @henriblancke 
I found the issue!
it's kind of silly, but the postgres JDBC Driver is expecting the key entry in the pkcs12 to have a [hardcoded alias "user"](https://github.com/pgjdbc/pgjdbc/blob/17e37a3cb4a7790f56481f08df88d579e77c6129/pgjdbc/src/main/java/org/postgresql/ssl/PKCS12KeyManager.java#L155). 
It's weird because their manager has code to get the first loaded alias in the store, but it doesn't seem to be used.

Anyway, putting "user" instead of the alias we have today makes the error go away
<img width="984" alt="image" src="https://user-images.githubusercontent.com/867491/193350783-aa77138f-8b81-43bf-825a-40164bbf645a.png">

**Comment by henriblancke** [comment]
@rodireich wow! awesome :) great find!! Thank you for all the help!

**Comment by rodireich** [comment]
Until we publish a fix, if you're running a local cluster (not cloud), you can set the value of io.airbyte.db.util.SSLCertificateUtils.KEYSTORE_ENTRY_PREFIX to unblock yourself:
`public static final String KEYSTORE_ENTRY_PREFIX = "user";`

**Comment by marcosmarxm** [comment]
@sebasaeros Rodi is working in a fix. If you want you can follow Rodi's suggestion and re-deploy the connector changing that variable.

**Comment by sebasaeros** [comment]
Now i am getting this error:

![image](https://user-images.githubusercontent.com/104517224/193642961-70885116-8791-4e1d-bac5-a7663e75eba6.png)

**Comment by sebasaeros** [comment]
Hello, thanks for the response,

which should be the value of the env variables public static final String KEYSTORE_ENTRY_PREFIX ?

My username?

Thanks

**Comment by sebasaeros** [comment]
@rodireich its ok thanks for the explanation.

I already have the -----BEGIN PRIVATE KEY----- statement, in which format need to be the key?

**Comment by rodireich** [comment]
It's whatever key you have. Can you share information on where you acquired this key from? 

Also, can you please check for validity with `openssl rsa -in /path/to/you/key -check` 
Do you get "RSA key ok"?
(make sure to not post private key here!)

**Comment by sebasaeros** [comment]
Yes, me database provider give me the key, I have the ca certificate, the user certificate and the user key

Key seem to be ok:

RSA key ok
writing RSA key
-----BEGIN RSA PRIVATE KEY-----

**Comment by rodireich** [comment]
@henriblancke and @sebasaeros 
A fixed version was deployed last night. You can give it a try if using OSS.
Or 1.0.14 should appear on cloud sometime today, depend on where you're using it.

@sebasaeros The key seems to be just an ordinary RSA key. Not sure why you're seeing this error.
Can you please share information on which db provider you're using? I'd like to understand if there's something different they are doing.

**Comment by henriblancke** [comment]
@rodireich thanks again for all the help!

**Comment by sebasaeros** [comment]
@rodireich everything is working now with the fix! 

Thanks for you help!
